### ISSUE TITLE
Potential Evaluation Issue: Binary Conversion of Multi-Class Problem May Lead to Inflated Metrics

### ISSUE THREAD

**Issue by glmanhtu** [Event-AHU/OpenPAR#58]
Dear authors,

Thank you for the interesting work and for sharing your code.

While reviewing your evaluation methodology, I noticed a potential issue in how the performance metrics (e.g., F1-score and accuracy) are computed. Specifically, it seems that the multi-class classification problem has been converted into a binary classification task per class, and then metrics are averaged across classes or groups.

While this may appear reasonable at first glance, it can significantly distort the results. For example:

Suppose the true class is 5 (out of 8 total classes), and the model predicts class 4.

In a standard multi-class F1 evaluation, this would be counted as a complete miss (i.e., F1 = 0 for this sample).

However, in the binarized setting, this sample may be evaluated as 7 correct (all negatives) and 1 incorrect (the single false negative or false positive), leading to a misleadingly high accuracy or F1 score.

Even if per-group averaging is applied afterward, this core issue remains and can lead to overestimating the model's performance.

Please let me know if I misunderstood the implementation or the reasoning behind this approach 

I have made the simple code to demonstrate this issue:

```python
# Sample data
#  [top color, age, gender, hat]
actual = np.array([7, 3, 1, 1])
pred = np.array([3, 2, 1, 1])

# Number of categories:
n_classes = [8, 4, 1, 1]

origin_f1 = f1_score(actual, pred, average='macro') # See https://github.com/yuange250/video_pedestrian_attributes_recognition/blob/master/main_video_attr_recog.py#L314
print(f"[Chen et al] F1: {origin_f1}")

f1_scores = []
for i, n_class in enumerate(n_classes):
    is_single = n_class == 1
    if not is_single:
        gt_one_hot = np.zeros((1, n_class))
        gt_one_hot[:, actual[i]] = 1
        
        pred_one_hot = np.zeros((1, n_class))
        pred_one_hot[:, pred[i]] = 1
        res = simple_par_metrics(gt_one_hot, pred_one_hot)  # See https://github.com/Event-AHU/OpenPAR/blob/main/EventPAR_Benchmark/VRWKV_PAR_Backbone/tools/function.py#L179
    else:
        single_gt = np.array([[actual[i]]])
        single_pred = np.array([[pred[i]]])
        res = simple_par_metrics(single_gt, single_pred, signle=True)
    f1_scores.append(res.f1)
    
f1_scores = np.mean(f1_scores)
print(f"[VTFPAR++] F1: {f1_scores.mean()}")
```

Output:
```bash
[Chen et al] F1: 0.25
[VTFPAR++] F1: 0.5
```

**Comment by NOP1224** [comment]
We previously did not notice the calculation method used by Chen et al. In their code, they seem to provide two types of results, macro and micro (#Line 329–330). The former (macro) treats multi-label classification as multiple groups of single-label classification, which clearly does not align with the conventional multi-label classification paradigm in the PAR field and introduces a strong prior assumption: each group contains only one attribute. This assumption oversimplifies the problem and does not match real-world conditions. The latter (micro) is consistent with our calculation method. If you have any doubts, please contact Chen to confirm which metric was reported in their paper. Below is our verification.
'''
import numpy as np
from sklearn.metrics import f1_score
from easydict import EasyDict

y_true_idx = np.array([7, 3, 1, 1])
y_pred_idx = np.array([3, 2, 1, 1])

def simple_par_metrics(gt_label, preds_probs, threshold=0.45, signle=False):
    pred_label = preds_probs > threshold
        
    eps = 1e-20
    result = EasyDict()
    if signle:
        preds_probs = preds_probs[:,np.newaxis]
    # TP
    TP = np.sum((gt_label == 1) * (pred_label == 1), axis=0).astype(float) 
    # TN
    TN = np.sum((gt_label == 0) * (pred_label == 0), axis=0).astype(float)
    # FP
    FP = np.sum(((gt_label == 0) * (pred_label == 1)), axis=0).astype(float)
    # FN
    FN = np.sum(((gt_label == 1) * (pred_label == 0)), axis=0).astype(float)

    TP =  np.sum(TP).astype(float)
    TN =  np.sum(TN).astype(float)
    FP =  np.sum(FP).astype(float)
    FN =  np.sum(FN).astype(float)

    acc = (TP + TN) / (TP + TN + FP + FN + eps)
    prec = (TP) / (TP + FP + eps)
    recall = (TP) / (TP + FN + eps)
    f1 = 2 * prec * recall / (prec + recall + eps)
    
    result.acc = acc
    result.prec = prec
    result.recall = recall
    result.f1 = f1

    return result

n_classes = [8, 4, 1, 1]
f1_scores = []
for i, n_class in enumerate(n_classes):
    is_single = n_class == 1
    if not is_single:
        gt_one_hot = np.zeros((1, n_class))
        gt_one_hot[:, y_true_idx[i]] = 1
        
        pred_one_hot = np.zeros((1, n_class))
        pred_one_hot[:, y_pred_idx[i]] = 1
        res = simple_par_metrics(gt_one_hot, pred_one_hot)  # See https://github.com/Event-AHU/OpenPAR/blob/dd89b94e0bf541d30214d1a509ccf33562870369/VTFPAR%2B%2B/tools/function.py#L177
    else:
        single_gt = np.array([[y_true_idx[i]]])
        single_pred = np.array([[y_pred_idx[i]]])
        res = simple_par_metrics(single_gt, single_pred, signle=True)
    f1_scores.append(res.f1)

print("(macro):", f1_score(y_true_idx, y_pred_idx, average="macro"))
print("(micro):", f1_score(y_true_idx, y_pred_idx, average="micro"))

print("binary:", np.mean(f1_scores))
'''
Result:
'''
(macro): 0.25
(micro): 0.5
binary: 0.5
‘’‘

**Comment by glmanhtu** [comment]
Hello @NOP1224,

Thank you for your clarification, 

Regarding the calculation results provided by Chen et al, since they treated the problem as a classification problem, for each attribute there will be only one class output (see https://github.com/yuange250/video_pedestrian_attributes_recognition/blob/master/main_video_attr_recog.py#L304). Then, in this case, f1_micro would be equal to accuracy, which is not the case according to their paper. Furthermore, according to their evaluation code (see https://github.com/yuange250/video_pedestrian_attributes_recognition/blob/8b2ac3130c04d43c87d02700c749bd95ce96bf4e/main_video_attr_recog.py#L314-L318), only macro f1 is being reported per attribute. 

But let's put that aside, accuracy calculation is also affected by this issue. 

```python

# Sample data
#  [top color, age, gender, hat]
actual = np.array([7, 3, 1, 1])
pred = np.array([3, 2, 1, 1])

# Number of categories:
n_classes = [8, 4, 1, 1]

origin_acc= np.sum(actual == pred)      # See https://github.com/yuange250/video_pedestrian_attributes_recognition/blob/master/main_video_attr_recog.py#L308
origin_acc /= len(actual)
print(f"[Chen et al] Acc: {origin_acc}")

acc_scores = []
for i, n_class in enumerate(n_classes):
    is_single = n_class == 1
    if not is_single:
        gt_one_hot = np.zeros((1, n_class))
        gt_one_hot[:, actual[i]] = 1
        
        pred_one_hot = np.zeros((1, n_class))
        pred_one_hot[:, pred[i]] = 1
        res = simple_par_metrics(gt_one_hot, pred_one_hot)  # See https://github.com/Event-AHU/OpenPAR/blob/dd89b94e0bf541d30214d1a509ccf33562870369/VTFPAR%2B%2B/tools/function.py#L177
    else:
        single_gt = np.array([[actual[i]]])
        single_pred = np.array([[pred[i]]])
        res = simple_par_metrics(single_gt, single_pred, signle=True)
    acc_scores.append(res.acc)

acc_scores = np.mean(acc_scores)
print(f"[VTFPAR++] Acc: {acc_scores.mean()}")
```
```Bash
Results:

[Chen et al] Acc: 0.5
[VTFPAR++] Acc: 0.8125
```

I understand the intention of your evaluation protocol and I agree that the evaluation protocol of Chen et al does not seem to match real-world conditions, e.g. there are many more top colors than the listed categories. Yet, since you are the one who follows their work and compare directly the performance of your work with theirs, it is mandatory that you have to understand their evaluation protocol and have a fair comparison, then once you have proven that your method is better then you can introduce a new metric or evaluation protocol as your wish.

**Comment by NOP1224** [comment]
Thank you for your valuable suggestions. We have re-evaluated our model using the criteria of **Chen et al**. The specific codes are provided in [train_marco.py](VTFPAR++/train_marco.py), [batch_engine_marco.py](VTFPAR++/batch_engine_marco.py), [base_block.py](VTFPAR++/models/base_block_marco.py). The final results are as follows:



```
Accs: [0.95373356, 0.94430662, 0.84185066, 0.91342099, 0.97928554, 0.89890846,
0.89580749, 0.94802778, 0.95621434, 0.75328703, 0.9461672, 0.78193997,
0.721161, 0.86913917]

Macro-F1: [0.7584219209705649, 0.9411434132068677, 0.7701404788385028, 0.88711170598405,
0.8318831762928199, 0.7867295109277703, 0.8822454777894155, 0.9460569264738823,
0.8945523666846384, 0.6008841912958576, 0.49172357039979875, 0.7616731698773676,
0.5536107056049715, 0.5334723207116681]

Average Accuracy: 0.8859, Average Macro-F1: 0.7600
```
It still remains at the leading position, and we only use the _**CrossEntropy**_ loss. We will gradually update the results in the paper in the future.

[base_block_marco.py](https://github.com/user-attachments/files/22065789/base_block_marco.py)
[batch_engine_marco.py](https://github.com/user-attachments/files/22065790/batch_engine_marco.py)
[train_marco.py](https://github.com/user-attachments/files/22065788/train_marco.py)

**Comment by glmanhtu** [comment]
Great, thank you for addressing the issue so quickly.
